### PR TITLE
refactor: migrate alert_configuration resource and removing deprecated fields

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Ticket: [INTMDB-996](https://jira.mongodb.org/browse/INTMDB-996)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments

- All acceptance tests are running, including the ones that require external credentials:
Screenshot 2023-08-16 at 10 35 33
- I encountered some scenarios of failing non-empty plans in migrations tests, more details here: https://discuss.hashicorp.com/t/terraform-framework-migration-tests-with-non-empty-plan-due-to-optional-attributes/57305. (update) **Solution**:  the optional attributes that could not be defined to null properly were set as computed, given that the api always responds with default values for them.

### CODE REVIEW

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go [hunk 1]
```diff
@@ -0,0 +1,811 @@
...
+		return
+	}
+
+	ids := decodeStateID(alertConfigState.ID.ValueString())
+
+	alert, getResp, err := conn.AlertConfigurations.GetAnAlertConfig(context.Background(), ids[encodedIDKeyProjectID], ids[encodedIDKeyAlertID])
+	if err != nil {
+		// deleted in the backend case
+		if getResp != nil && getResp.StatusCode == http.StatusNotFound {
+			alertConfigState.ID = types.StringValue("") // TODO verify this works
```
Comment (AgustinBettati) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: need to verify this works the same why it does in sdk v2 for marking resource as destroyed

Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: fixed to correct usage of remove resource method.

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go [hunk 2]
```diff
@@ -60,13 +59,8 @@ func TestAccConfigRSAlertConfiguration_EmptyMetricThresholdConfig(t *testing.T)
...
 				Config: testAccMongoDBAtlasAlertConfigurationConfigEmptyMetricThresholdConfig(orgID, projectName, true),
-				Check: resource.ComposeTestCheckFunc(
-					testAccCheckMongoDBAtlasAlertConfigurationExists(resourceName, alert),
-					resource.TestCheckResourceAttrSet(resourceName, "project_id"),
-					resource.TestCheckResourceAttr(resourceName, "notification.#", "1"),
-					resource.TestCheckResourceAttrSet(resourceName, "threshold_config.#"),
-				),
-				ExpectNonEmptyPlan: true,
+				// metric threshold is not created so state is inconsistent with plan
+				ExpectError: regexp.MustCompile("Error: Provider produced inconsistent result after apply"),
```
Comment (AgustinBettati) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go:63]: When an empty metric threshold config block is defined, sdk v2 would provide a non empty plan showing differences after the apply, while framework now directly throws an error display inconsistent state after apply.

File: mongodbatlas/data_source_mongodbatlas_alert_configuration.go [hunk 3]
```diff
@@ -504,3 +507,178 @@ func convertNotificationToCtyValues(notification *matlas.Notification) map[strin
 
 	return values
 }
+
```
Comment (AgustinBettati) [inline mongodbatlas/data_source_mongodbatlas_alert_configuration.go:510]: these are methods that I moved from old resource implementation as they are used in the data source, will be removed when data source is migrated as well.

File: mongodbatlas/data_source_mongodbatlas_alert_configuration.go [hunk 4]
```diff
@@ -504,3 +507,178 @@ func convertNotificationToCtyValues(notification *matlas.Notification) map[strin
...
+			"metric_name": m.MetricName,
+			"operator":    m.Operator,
+			"threshold":   cast.ToString(m.Threshold),
+			"units":       m.Units,
+			"mode":        m.Mode,
+		}
+	}
+
+	return map[string]interface{}{}
+}
```
Comment (andreaangiolillo) [inline mongodbatlas/data_source_mongodbatlas_alert_configuration.go:523]: why are you still using `map[string]interface` instead of go struct?

File: mongodbatlas/data_source_mongodbatlas_alert_configuration.go [hunk 5]
```diff
@@ -504,3 +507,178 @@ func convertNotificationToCtyValues(notification *matlas.Notification) map[strin
...
+			"operator":  m.Operator,
+			"units":     m.Units,
+			"threshold": m.Threshold,
+		}}
+	}
+
+	return []interface{}{}
+}
+
+func expandAlertConfigurationNotification(d *schema.ResourceData) ([]matlas.Notification, error) {
```
Comment (andreaangiolillo) [inline mongodbatlas/data_source_mongodbatlas_alert_configuration.go:563]: instead of using `expand**` let's use go convention `newMongoDBAtlasNotification` (`new[NameOfStructThatIsReturned`)

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go [hunk 6]
```diff
@@ -0,0 +1,811 @@
...
+	alertConfigurationID := parts[1]
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("id"), encodeStateID(map[string]string{
+		"id":         alertConfigurationID,
+		"project_id": projectID,
+	}))...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("project_id"), projectID)...)
+}
+
+func expandTFAlertConfigurationNotification(tfNotificationSlice []tfNotificationModel) ([]matlas.Notification, error) {
```
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: same as above. I would like to move away from expand - flat format and use `new[NameOfResource]` which is widely used in golang

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go [hunk 7]
```diff
@@ -0,0 +1,811 @@
...
+				OpsGenieRegion: conversion.StringValueToFramework(value.OpsGenieRegion),
+				TeamID:         conversion.StringValueToFramework(value.TeamID),
+				TypeName:       conversion.StringValueToFramework(value.TypeName),
+				Username:       conversion.StringValueToFramework(value.Username),
+				EmailEnabled:   types.BoolValue(value.EmailEnabled != nil && *value.EmailEnabled),
+				SMSEnabled:     types.BoolValue(value.SMSEnabled != nil && *value.SMSEnabled),
+			}
+		}
+	} else {
+		for i := range matlasSlice {
```
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]:
```suggestion
	if len(matlasSlice) != len(currStateNotifications) { // notifications were modified elsewhere from terraform, or import statement is being called
		for i := range matlasSlice {
			value := matlasSlice[i]
			notifications[i] = tfNotificationModel{
				TeamName:       types.StringValue(value.TeamName),
				Roles:          value.Roles,
				ChannelName:    conversion.StringValueToFramework(value.ChannelName),
				DatadogRegion:  conversion.StringValueToFramework(value.DatadogRegion),
				DelayMin:       types.Int64Value(int64(*value.DelayMin)),
				EmailAddress:   conversion.StringValueToFramework(value.EmailAddress),
				IntervalMin:    types.Int64Value(int64(value.IntervalMin)),
				MobileNumber:   conversion.StringValueToFramework(value.MobileNumber),
				OpsGenieRegion: conversion.StringValueToFramework(value.OpsGenieRegion),
				TeamID:         conversion.StringValueToFramework(value.TeamID),
				TypeName:       conversion.StringValueToFramework(value.TypeName),
				Username:       conversion.StringValueToFramework(value.Username),
				EmailEnabled:   types.BoolValue(value.EmailEnabled != nil && *value.EmailEnabled),
				SMSEnabled:     types.BoolValue(value.SMSEnabled != nil && *value.SMSEnabled),
			}
			return notifications
		}
		for i := range matlasSlice {
```

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go [hunk 8]
```diff
@@ -0,0 +1,811 @@
...
+	matchers := make([]tfMatcherModel, len(matlasSlice))
+	if len(matlasSlice) != len(currStateSlice) { // matchers were modified elsewhere from terraform, or import statement is being called
+		for i, value := range matlasSlice {
+			matchers[i] = tfMatcherModel{
+				FieldName: conversion.StringValueToFramework(value.FieldName),
+				Operator:  conversion.StringValueToFramework(value.Operator),
+				Value:     conversion.StringValueToFramework(value.Value),
+			}
+		}
+	} else {
```
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]:
```suggestion
	if len(matlasSlice) != len(currStateSlice) { // matchers were modified elsewhere from terraform, or import statement is being called
		for i, value := range matlasSlice {
			matchers[i] = tfMatcherModel{
				FieldName: conversion.StringValueToFramework(value.FieldName),
				Operator:  conversion.StringValueToFramework(value.Operator),
				Value:     conversion.StringValueToFramework(value.Value),
			}
		}
		return matches
```

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"mongodbatlas": {
+						VersionConstraint: "1.11.0",
+						Source:            "mongodb/mongodbatlas",
+					},
+				},
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckMongoDBAtlasAlertConfigurationExists(resourceName, alert),
+					resource.TestCheckResourceAttrSet(resourceName, "project_id"),
```
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go:78]: can we check for additional fields?

File: mongodbatlas/framework/conversion/string.go [hunk 10]
```diff
@@ -17,3 +17,12 @@ func TypesListToString(ctx context.Context, set types.List) []string {
 	_ = set.ElementsAs(ctx, &results, false)
 	return results
 }
+
+// StringValueToFramework converts a string value to a Framework String value.
+// An empty string is converted to a null String. Useful for optional attributes.
+func StringValueToFramework(v string) types.String {
+	if v == "" {
+		return types.StringNull()
```
Comment (andreaangiolillo) [inline mongodbatlas/framework/conversion/string.go:25]: why we cannot use types.StringValue(v) when v is `""`?

Comment (maastha) [reply mongodbatlas/framework/conversion/string.go:25]: I believe that's because in SDKv2 there was no concept of null values for string, ints etc so they always stored "" (for strings) or 0 (for integers) whether or not the user provided value for that attribute in the config. 
But now the Framework expects the state to be same as config and if the user has not specified an attribute in the config it is expected to be null in the state (sometimes we get "" for some values from the API, so Terraform throws inconsistent values error)
@AgustinBettati correct me if I'm wrong.

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+						"operator": schema.StringAttribute{
+							Optional: true,
+						},
+						"threshold": schema.Float64Attribute{
+							Optional: true,
+						},
+						"units": schema.StringAttribute{
+							Optional: true,
+							Validators: []validator.String{
+								stringvalidator.OneOf(
```
Comment (maastha) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go:231]: Does this validator ignoreCase as in validation.StringInSlice()? See previous [implementation](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/mongodbatlas/resource_mongodbatlas_alert_configuration.go#L155)
Applicable to all other uses here

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 6 above)
Comment (maastha) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: +1

File: mongodbatlas/framework/conversion/string.go [hunk 12]
```diff
@@ -17,3 +17,12 @@ func TypesListToString(ctx context.Context, set types.List) []string {
 	_ = set.ElementsAs(ctx, &results, false)
 	return results
 }
+
+// StringValueToFramework converts a string value to a Framework String value.
+// An empty string is converted to a null String. Useful for optional attributes.
+func StringValueToFramework(v string) types.String {
```
Comment (maastha) [inline mongodbatlas/framework/conversion/string.go]:
```suggestion
func StringNullIfEmpty(v string) types.String {
```

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
+package mongodbatlas
```
Comment (maastha) [inline mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go:1]: Can we add test cases for each collection in the resource (empty/null if optional, multiple elements wherever possible)
I have had to add logic for null collections, hence we should make sure

File: mongodbatlas/framework/conversion/string.go (hunk 10 above)
Comment (AgustinBettati) [reply mongodbatlas/framework/conversion/string.go:25]: Correct, this is relevant primarily for migration tests. If the empty string is defined directly you end up with non-empty plan diffs (related issue: https://github.com/hashicorp/terraform/issues/31887).

File: mongodbatlas/data_source_mongodbatlas_alert_configuration.go (hunk 5 above)
Comment (AgustinBettati) [reply mongodbatlas/data_source_mongodbatlas_alert_configuration.go:563]: you can ignore additions to the data_source_mongodbatlas_alert_configuration, these are legacy methods I moved from resource file to data source as they were being used, but when migrating the data sources all of these functions will be removed/replaced.

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 6 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: agree to moving forward with this convention. Just to be clear, for the case of converting from api models to terraform models I see that the toTF* prefix is being used (in project resource and my changes). Are we in agreement to rename these to newTF*  as well?

File: mongodbatlas/data_source_mongodbatlas_alert_configuration.go (hunk 4 above)
Comment (AgustinBettati) [reply mongodbatlas/data_source_mongodbatlas_alert_configuration.go:523]: more detail in comment below

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 11 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go:231]: Previous implementation defines `ignoreCase` as false for all StringInSlice calls, meaning it is case sensitive. I have verified that stringvalidator.OneOf() is also caes sensitive, so we should be good here.

File: mongodbatlas/data_source_mongodbatlas_alert_configuration.go (hunk 5 above)
Comment (andreaangiolillo) [reply mongodbatlas/data_source_mongodbatlas_alert_configuration.go:563]: ah, I see. Thanks for the clarification here

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 6 above)
Comment (andreaangiolillo) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: my vote is for `newTF*` 👍

File: mongodbatlas/framework/conversion/string.go (hunk 10 above)
Comment (andreaangiolillo) [reply mongodbatlas/framework/conversion/string.go:25]: Thanks for the clarifications here

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go (hunk 13 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go:1]: Added additional migration tests to verify all scenarios:
- empty/multiple matchers
- empty/defined metric threshold config
- empty/defined threshold config
- multiple notifications

File: mongodbatlas/framework/conversion/string.go (hunk 12 above)
Comment (AgustinBettati) [reply mongodbatlas/framework/conversion/string.go]: agree, adjusted

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 6 above)
Comment (maastha) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: I find `toTF*` more readable, but no strong opinion, whatever works for the team.

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go (hunk 2 above)
Comment (maastha) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go:63]: Are we planning to handle this error? Or are we waiting for a response from Hashicorp on this? 
Migration tests are basically to ensure backward compatibility so this would be a breaking change.

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 11 above)
Comment (maastha) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go:231]: Great, thank you!

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go (hunk 6 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go]: adjusted.

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go (hunk 2 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go:63]: This acceptance test defines an an empty block `metric_threshold_config {}` which does not have an initial validation as as all attributes are optional. 
- In 1.11.0 the apply runs successfully, but in the terraform state `metric_threshold_config` is an empty array. That is why when running terraform plan afterwards you receive a non empty plan as terraform mentions that this block needs to be created.
- With the new framework, after running terraform apply the output state is the same, with `metric_threshold_config` as an empty array. However the terraform apply responds with an error because the framework now validates that the state is inconsistent with the defined configuration.

I am not that concerned about this case as it is essentially an incorrectly defined configuration that users would fix when using 1.11.0

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go (hunk 9 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration_migration_test.go:78]: added additional checks for all migrations tests 👍

File: mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go (hunk 2 above)
Comment (maastha) [reply mongodbatlas/fw_resource_mongodbatlas_alert_configuration_test.go:63]: Got it, makes sense now, so this essentially is an invalid test case. I thought this was about a null list but I see it is actually defined as empty.
In my case, there were optional lists which even if not defined in the config at all were being saved as empty arrays in the state.
Thanks for the clarification, Agustin!

**Review by andreaangiolillo — APPROVED** [review]
LGTM. Thanks for addressing all the comments

**Review by maastha — APPROVED** [review]
LGTM
